local-cluster/tests/local_cluster.rs

```diff
@@ -1680,7 +1601,282 @@ fn test_validator_saves_tower() {
+        "4mx9yoFBeYasDKBGDWCTWGJdWuJCKbgqmuP8bN9umybCh5Jzngw7KQxe99Rf5uzfyzgba1i65rJW4Wqk7Ab5S8ye",
+        "3zsEPEDsjfEay7te9XqNjRTCE7vwuT6u4DHzBJC19yp7GS8BuNRMRjnpVrKCBzb3d44kxc4KPGSHkCmk6tEfswCg",
+    ]
+    .iter()
+    .map(|s| (Arc::new(Keypair::from_base58_string(s)), true))
+    .take(node_stakes.len())
+    .collect::<Vec<_>>();
+    let validators = validator_keys
+        .iter()
+        .map(|(kp, _)| kp.pubkey())
+        .collect::<Vec<_>>();
+    let (validator_a_pubkey, validator_b_pubkey, validator_c_pubkey) =
+        (validators[0], validators[1], validators[2]);
+
+    let config = ClusterConfig {
+        cluster_lamports: 100_000,
+        node_stakes: node_stakes.clone(),
+        validator_configs: vec![ValidatorConfig::default(); node_stakes.len()],
+        validator_keys: Some(validator_keys),
+        slots_per_epoch,
+        stakers_slot_offset: slots_per_epoch,
+        skip_warmup_slots: true,
+        ..ClusterConfig::default()
+    };
+    let mut cluster = LocalCluster::new(&config);
+
+    let base_slot = 26; // S2
+    let next_slot_on_a = 27; // S3
+    let truncated_slots = 100; // just enough to purge all following slots after the S2 and S3
+
+    // Immediately kill validator C
+    let validator_c_info = cluster.exit_node(&validator_c_pubkey);
+
+    // Step 1:
+    // Let validator A, B, (D) run for a while.
+    let client = cluster.get_validator_client(&validator_a_pubkey).unwrap();
+    loop {
+        sleep(Duration::from_millis(100));
+        let highest_bank = client
+            .get_slot_with_commitment(CommitmentConfig::recent())
+            .unwrap();
+        if highest_bank >= next_slot_on_a {
+            break;
+        }
+    }
+    let validator_a_info = cluster.exit_node(&validator_a_pubkey);
+    let client = cluster.get_validator_client(&validator_b_pubkey).unwrap();
+    loop {
+        sleep(Duration::from_millis(100));
+        let highest_bank = client
+            .get_slot_with_commitment(CommitmentConfig::recent())
+            .unwrap();
+        if highest_bank >= next_slot_on_a {
+            break;
+        }
+    }
+    let _validator_c_info = cluster.exit_node(&validator_b_pubkey);
+
+    // Step 2:
+    // Stop validator and truncate ledger
+    info!("truncate validator C's ledger");
+    {
+        // first copy from validator A's ledger
+        std::fs::remove_dir_all(&validator_c_info.info.ledger_path).unwrap();
+        let mut opt = fs_extra::dir::CopyOptions::new();
+        opt.copy_inside = true;
+        fs_extra::dir::copy(
+            &validator_a_info.info.ledger_path,
+            &validator_c_info.info.ledger_path,
+            &opt,
+        )
+        .unwrap();
+        std::fs::remove_file(Tower::get_filename(
+            &validator_c_info.info.ledger_path,
+            &validator_a_pubkey,
+        ))
+        .unwrap();
+
+        let blockstore = Blockstore::open_with_access_type(
```

Comment:
Maybe we can extract this open -> purge logic into a separate function because the same logic is reused below?

Reply:
done